### PR TITLE
Add Tcp{Client,Listener} and UdpClient SendReceive tests.

### PR DESCRIPTION
**Pull request by pgavlin**
This exercises NetworkStream as well as the aforementioned types.

### CODE REVIEW

**Comment by pgavlin** [comment]
@CIPop @davidsh @stephentoub PTAL

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs [hunk 1]
```diff
@@ -125,6 +125,77 @@ private static void SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAd
...
+                    Assert.Null(receivedChecksums[datagramId]);
+
+                    receivedChecksums[datagramId] = Fletcher32.Checksum(result.Buffer, 0, result.Buffer.Length);
+                }
+            });
+
+            var sentChecksums = new uint[DatagramsToSend];
+            int sentDatagrams = 0;
+
+            Task.Run(async () =>
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: We should wait for this task to complete somewhere.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs [hunk 2]
```diff
@@ -310,6 +381,74 @@ private static void SendRecvAsync_Stream_TCP(IPAddress listenAt, bool useMultipl
...
+
+                        bytesReceived += received;
+                        receivedChecksum.Add(recvBuffer, 0, received);
+                    }
+                }
+            });
+
+            int bytesSent = 0;
+            var sentChecksum = new Fletcher32();
+            Task.Run(async () =>
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: Ditto.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs [hunk 3]
```diff
@@ -125,6 +125,77 @@ private static void SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAd
...
+                    Assert.True(receiverAck.Wait(AckTimeout));
+                    receiverAck.Reset();
+                    senderAck.Set();
+
+                    Assert.Equal(DatagramSize, sent);
+                    sentChecksums[sentDatagrams] = Fletcher32.Checksum(sendBuffer, 0, sent);
+                }
+            });
+
+            Assert.True(receiverTask.Wait(TestTimeout));
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: You could make this a WaitAll to wait on both tasks with the one timeout.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs (hunk 1 above)
Comment (pgavlin) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: Done.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs (hunk 2 above)
Comment (pgavlin) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: Done.

**Comment by stephentoub** [comment]
Left a couple comments on waiting for tasks, in particular so that asserts propagate, but otherwise LGTM.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs [hunk 4]
```diff
@@ -125,6 +125,77 @@ private static void SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAd
...
+            var rightEndpoint = (IPEndPoint)right.Client.LocalEndPoint;
+
+            var receiverAck = new ManualResetEventSlim();
+            var senderAck = new ManualResetEventSlim();
+
+            var receivedChecksums = new uint?[DatagramsToSend];
+            int receivedDatagrams = 0;
+
+            var receiverTask = Task.Run(async () =>
+            {
```
Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: Should you use `var` here or use `Task` ?  Comment applies to other instances of this in the file.

Comment (pgavlin) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: Technically, these should be `Task`. Good catch.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs [hunk 5]
```diff
@@ -125,6 +125,77 @@ private static void SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAd
             }
         }
 
+        private static void SendToRecvFromAsync_UdpClient_Datagram_UDP(IPAddress leftAddress, IPAddress rightAddress)
+        {
```
Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: nit: Why use capital `UDP` in the test name but other parts of the test name use `Udp` ?

**Comment by davidsh** [comment]
left small comments, otherwise LGTM.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs (hunk 5 above)
Comment (pgavlin) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: One is a protocol name, one is part of the name of a type.

Comment (davidsh) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: ok

**Comment by pgavlin** [comment]
@dotnet-bot test this please

**Comment by pgavlin** [comment]
@dotnet-bot test this please

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs [hunk 6]
```diff
@@ -125,6 +125,77 @@ private static void SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAd
...
+                    Assert.Equal(DatagramSize, sent);
+                    sentChecksums[sentDatagrams] = Fletcher32.Checksum(sendBuffer, 0, sent);
+                }
+            });
+
+            Assert.True(Task.WaitAll(new[] { receiverTask, senderTask }, TestTimeout));
+            for (int i = 0; i < DatagramsToSend; i++)
+            {
+                Assert.NotNull(receivedChecksums[i]);
+                Assert.Equal(sentChecksums[i], (uint)receivedChecksums[i]);
```
Comment (CIPop) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: This should be usually true when using `loopback` and the extra locking above. 
An improvement would be to add a simple matching protocol and allow for missed receives.

To at least avoid getting people the wrong idea, we should add a clarifying comment that, if used over real networks (or in certain conditions on the same machine), receiving all datagrams or receiving ordered datagrams is not ensured by the UDP protocol.

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs [hunk 7]
```diff
@@ -125,6 +125,77 @@ private static void SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAd
             }
         }
 
+        private static void SendToRecvFromAsync_UdpClient_Datagram_UDP(IPAddress leftAddress, IPAddress rightAddress)
```
Comment (CIPop) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: Nit: for clarity we usually split the tests first by class, not async technology. I would place these tests in separate UdpClient / TcpClient test files.
One other reason is if we ever want to test only the 4.0 contract where these types don't exist (e.g. some security fix).

**Comment by CIPop** [comment]
Left a few comments about code organization and maybe adding some clarification on the UDP test.

LGTM

File: src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs (hunk 6 above)
Comment (pgavlin) [inline src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs]: > An improvement would be to add a simple matching protocol and allow for missed receives.

There is a matching protocol: the first byte of each datagram is a sequence number. There is not currently any accommodation for dropped packets, however. This is something I'm thinking about at the moment.

**Comment by pgavlin** [comment]
@dotnet-bot test this please

**Comment by pgavlin** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by pgavlin** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
Looks like the sockets tests crashed:
```
Discovering: System.Net.Sockets.Tests
  Discovered:  System.Net.Sockets.Tests
  Starting:    System.Net.Sockets.Tests
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00103\lib\tests.targets(129,5): warning MSB3073: The command "CoreRun.exe xunit.console.netcore.exe System.Net.Sockets.Tests.dll  -xml testResults.xml -notrait category=nonwindowstests  -notrait category=OuterLoop -notrait category=failing " exited with code -1073741819. [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\tests\FunctionalTests\System.Net.Sockets.Tests.csproj]
d:\j\workspace\dotnet_corefx_windows_release_prtest\packages\Microsoft.DotNet.BuildTools.1.0.25-prerelease-00103\lib\tests.targets(136,5): error : One or more tests failed while running tests from 'System.Net.Sockets.Tests' please check log for details! [d:\j\workspace\dotnet_corefx_windows_release_prtest\src\System.Net.Sockets\tests\FunctionalTests\System.Net.Sockets.Tests.csproj]
```

**Comment by pgavlin** [comment]
Yeah. Not sure why yet: this was fine on Windows earlier today (locally and in CI). Investigating now.

**Comment by pgavlin** [comment]
This appears to be fine now.

**Pull request merged**